### PR TITLE
Grid Summary Row

### PR DESCRIPTION
**Pull request by haynesjm42**
#737 

Support for showing a summary/total row in our Grid component pinned to the top or bottom (uses ag-Grid pinned row feature)

In order to use the summary row the application needs to include a row in the store with an id of "summary" which contains the data to display, and set the new `showSummary` prop on the GridModel. If the store contains hierarchical data, the "summary" record will be excluded from the main record hierarchy (it is stored separately in RecordSet). I went through several iterations on this implementation and landed on this as the best approach. In the future if we end up merging the Aggregation support in Cube/CubeRecord over to Store/Record the interface for the summary support shouldn't need to change and would be able to handle summary record without needing the application to provide the it.

A caveat of using the summary row is that any app-defined pinned rows will be overwritten when the Grid component updates the summary row. An alternative approach to achieving summary rows in ag-Grid would be to use aligned grids, but that seemed like it would be a more complex implementation for our Grid/GridModel, though it has the advantage of not conflicting with pinned rows an application may want to set.

### CODE REVIEW

**Comment by haynesjm42** [comment]
Toolbox example: https://github.com/exhi/toolbox/pull/196

File: data/Record.js [hunk 1]
```diff
@@ -52,6 +52,9 @@ export class Record {
         return this.store.getChildrenById(this.id, false);
     }
 
+    get isSummary() {
+        return this.id === 'summary';
```
Comment (lbwexler) [inline data/Record.js]: I don't think we want to require a special id.   The id's of our root nodes can be significant.  Let's discuss whether it would be ok to assume that a single row with zero siblings is the summary?

Comment (haynesjm42) [reply data/Record.js]: tl;dr - I ran into complications trying to use the shape of the data to identify the summary record, so went down the path of making it explicitly identified.

I was initially handling it the way you suggest, with treating a root node with no siblings as the "summary" record, but that really didn't work out well in all cases.

In the case that the data itself is not hierarchical I didn't want to force apps to add a root node as the means of supplying a summary record, though I could be persuaded on this point.

Also in order to make sure the summary record itself is handled correctly when filtering, I am keeping it completely separated from the main record list/map. This seemed like the cleanest solution for this, and didn't require that additional logic be added in various methods in RecordSet to handle the summary record. To have things work that way it is necessary to know exactly which record is the summary record when the data is processed by RecordSet, and since it is entirely possible that the data being loaded into the store has a single root node which is _not_ a summary record, it is necessary to be explicit when identifying the summary record instead of following some kind of convention based on the shape of the data.

I see a few options here for explicitly identifying the summary record:

1. Have a "special" id we use to identify, but make this something that can be configured on the Store (`summaryIdSpec`?). This would allow the app to keep their root ids. We could default it to `xhSummaryRecordId` or something and have a constant to clean up the code and make it easy for apps to access it. 
2. Instead of using the id, we could look for an `isSummary` prop on the raw data to identify the summary record. This would mean reserving another prop on Record that apps couldn't use for fields, and wanted to avoid doing that, so I went with the id approach.
3. Separate the summary data even more from the other data by having a `setSummaryData` or `loadSummaryData` on Store, and never even keep it with the other records. Would require either having a separate RecordSet for the summary record or some refactoring would be needed to be able to create the actual record without going through a RecordSet.

Please let me know if I am over-complicating things here :) Would be happy to have a simpler means of managing the summary record.

**Comment by haynesjm42** [comment]
The summary row should not be effected by store filtering at all, it is kept separate from the other records for this reason. The summary row will always be visible, even if the store filter would result in 0 records shown.

The same implementation should work fine on mobile, I added an example in the toolbox branch for the mobile tree grid. Were you thinking we needed a different approach for summary rows on mobile?

**Comment by lbwexler** [comment]
John and I went through this in great detail this afternoon.  In particular, he explained to me the rational behind having an explicit "out-of-band" handling of the summary record in Store, vs. just interpreting the single root as the total.

I have to say I wasn't 100% convinced and we agreed to continue thinking about it and get others to chime in.  I am still kind of partial for just looking for a root of any kind.  John mentioned that in the case of flat data, it's strange to introduce the hierarchical data/children thing for the purposes of the total row alone.

It does seem to me that we would never want the filter to whack our total row, so we want storeFilter to also recognize the summary in some way.....

In any case,  I do feel pretty strongly that if we want to identify a total row in the rawData other than having it be the root,  I would prefer a dedicated flag along the lines of "isStoreTotal' versus putting the magic string in the id....

File: data/Store.js [hunk 2]
```diff
@@ -74,25 +83,62 @@ export class Store {
...
-    loadData(rawData) {
+    loadData(rawData, rawSummaryData) {
+        const {_treatRootAsSummary} = this;
+        throwIf(_treatRootAsSummary && rawSummaryData,
+            'Store is configured to treat the root node as summary data but rawSummaryData was provided to loadData!');
+
+        if (this.haveRootSummary(rawData)) {
+            rawSummaryData = rawData[0];
+            rawData = rawSummaryData.children;
+            delete rawSummaryData.children;
```
Comment (lbwexler) [inline data/Store.js]: hmm -- we were trying not to mutate the raw data in this object.  Shouldn't we clone the rawData[0] here?

File: data/cube/impl/QueryExecutor.js [hunk 3]
```diff
@@ -20,7 +16,7 @@ export class QueryExecutor {
     static async getDataAsync(query) {
         const {dimensions, includeRoot, fields, cube, filters} = query,
             cubeRecords = cube.records,
```
Comment (lbwexler) [inline data/cube/impl/QueryExecutor.js]: should be able to revert this change
